src/backend/concurrency/eager_write_txn_manager.cpp

```diff
@@ -0,0 +1,672 @@
+  assert(IsOwner(tile_group_header, tuple_id) == false);
+
+  // Try to get write lock
+  auto current_tid = current_txn->GetTransactionId();
+  auto old_tid = INITIAL_TXN_ID;
+
+  GetEwReaderLock(tile_group_header, tuple_id);
+
+  old_tid =
+      tile_group_header->SetAtomicTransactionId(tuple_id, old_tid, current_tid);
+
+  // Check if we successfully get the write lock
+  bool res = (old_tid == INITIAL_TXN_ID);
+  if (!res) {
+    LOG_INFO("Fail to acquire write lock. Set txn failure.");
+    ReleaseEwReaderLock(tile_group_header, tuple_id);
+    return false;
+  }
+
+  // Install wait for dependency on all reader txn
+  auto ptr = GetEwReaderList(tile_group_header, tuple_id);
+  while (ptr->next != nullptr) {
+    // try to install a wait for in every reader
+    auto reader_tid = ptr->next->txn_id_;
+    // skip my self
+    if (reader_tid == current_tid) {
+      ptr = ptr->next;
+      continue;
+    }
+
+    // Lock the txn bucket
+    {
+      std::lock_guard<std::mutex> lock(running_txn_map_mutex_);
+      if (running_txn_map_.count(reader_tid) != 0) {
+        // Add myself to the wait_for set of reader
+        auto reader_ctx = running_txn_map_[reader_tid];
+        LOG_INFO("Add dependency to %lu", reader_tid);
+        if (reader_ctx->wait_list_.insert(current_tid).second == true) {
+          // New dependency
+          current_txn_ctx->wait_for_counter_++;
+        }
+      } else {
+        assert(false);
+      }
+    }
+    ptr = ptr->next;
+  }
+
+  ReleaseEwReaderLock(tile_group_header, tuple_id);
+  // Atomically increase the wait_for counter
+  assert(current_txn_ctx->wait_for_counter_ >= 0);
+
+  return true;
+}
+
+
+// release write lock on a tuple.
+// one example usage of this method is when a tuple is acquired, but operation
+// (insert,update,delete) can't proceed, the executor needs to yield the 
+// ownership before return false to upper layer.
+// It should not be called if the tuple is in the write set as commit and abort
+// will release the write lock anyway.
+// TODO: may want to do other stuffs here
+void EagerWriteTxnManager::YieldOwnership(const oid_t &tile_group_id,
+  const oid_t &tuple_id) {
+
+  auto &manager = catalog::Manager::GetInstance();
+  auto tile_group_header = manager.GetTileGroup(tile_group_id)->GetHeader();
+  assert(IsOwner(tile_group_header, tuple_id));
+  tile_group_header->SetTransactionId(tuple_id, INITIAL_TXN_ID);
+}
+
+bool EagerWriteTxnManager::PerformRead(const ItemPointer &location) {
+  oid_t tile_group_id = location.block;
+  oid_t tuple_id = location.offset;
+
+  LOG_INFO("Perform read");
+  auto &manager = catalog::Manager::GetInstance();
+  auto tile_group = manager.GetTileGroup(tile_group_id);
+  auto tile_group_header = tile_group->GetHeader();
+
+  auto &rw_set = current_txn->GetRWSet();
+  auto tuple_map = rw_set.find(tile_group_id);
+  if (tuple_map != rw_set.end()) {
+    if (tuple_map->second.find(tuple_id) != tuple_map->second.end()) {
```

Comment:
These two conditions could be combined into the same if statement with the `&&` operator.